File: WebClient SDK/src/default/ui/krProcess/stageHandlers/tabCaptionUIHandler.ts
import { IKrStageTypeUIHandlerContext, KrStageTypeUIHandler } from 'tessa/workflow/krProcess';

import { MetadataStorage } from 'tessa';
import { TabContentIndicator } from '../../tabContentIndicator';
import { TabControlViewModel } from 'tessa/ui/cards/controls';

/**
 * UI обработчик этапов, настраивающий заголовки стандартных вкладок на форме с настройками этапа: Условие, Инициализация, Постобработка.
 */
export class TabCaptionUIHandler extends KrStageTypeUIHandler {
  //#region fields

  private _dispose: Function | null;

  //#endregion

  //#region base overrides

  public async initialize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    const control = context.rowModel.controls.get('CSharpSourceTable');
    if (control && control instanceof TabControlViewModel) {
      const sectionMeta = MetadataStorage.instance.cardMetadata.getSectionByName('KrStagesVirtual');

      if (!sectionMeta) {
        return;
      }

      const fieldIds: [string, string][] = sectionMeta.columns.map(x => [x.id || '', x.name || '']);

      const indicator = new TabContentIndicator(control, context.row.getStorage(), fieldIds, true);
      indicator.update();
      this._dispose = context.row.fieldChanged.addWithDispose(indicator.fieldChangedAction);
    }
  }

  public async finalize(): Promise<void> {
    if (this._dispose) {
      this._dispose();
      this._dispose = null;
    }
  }

  //#endregion
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/StageHandlers/TabCaptionUIHandler.cs
#nullable enable

using System.Linq;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Default.Client.UI.KrProcess.StageHandlers
{
    /// <summary>
    /// UI обработчик этапов, настраивающий заголовки стандартных вкладок на форме с настройками этапа: Условие, Инициализация, Постобработка.
    /// </summary>
    public sealed class TabCaptionUIHandler :
        StageTypeUIHandlerBase
    {
        #region Fields

        private TabContentIndicator? indicator;

        #endregion

        #region Base Overrides

        /// <inheritdoc />
        public override async Task Initialize(
            IKrStageTypeUIHandlerContext context)
        {
            if (context.RowModel.Controls.TryGet(KrConstants.Ui.CSharpSourceTable, out var control)
                && control is TabControlViewModel tabControl)
            {
                var sectionMeta = (await context.CardModel.CardMetadata.GetSectionsAsync(context.CancellationToken).ConfigureAwait(false))[KrConstants.KrStages.Virtual];
                var fieldIDs = sectionMeta.Columns.ToDictionary(k => k.ID, v => v.Name);

                this.indicator = new TabContentIndicator(tabControl, context.Row, fieldIDs, true);
            }
        }

        /// <inheritdoc />
        public override Task Finalize(
            IKrStageTypeUIHandlerContext context)
        {
            if (this.indicator is not null)
            {
                this.indicator.Dispose();
                this.indicator = null;
            }

            return Task.CompletedTask;
        }

        #endregion
    }
}
